### PR TITLE
GossipSub: IDontWant

### PR DESCRIPTION
**Pull request by Menduist**
Implements https://github.com/libp2p/specs/pull/548

### CODE REVIEW

**Comment by Menduist** [comment]
Confirmed to work properly in sims (~15% BW reduction with: 500 nodes, 100mbps, 100ms ping, 0.5kb messages)
[Stagger sending](https://github.com/status-im/nim-libp2p/issues/850) will reduce the bandwidth even further, since we will have more time to avoid a duplicate

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 1]
```diff
@@ -332,11 +333,25 @@ proc validateAndRelay(g: GossipSub,
       g.floodsub.withValue(t, peers): toSendPeers.incl(peers[])
       g.mesh.withValue(t, peers): toSendPeers.incl(peers[])
 
+
```
Comment (diegomrsantos) [inline libp2p/protocols/pubsub/gossipsub.nim:336]: not necessary?

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 2]
```diff
@@ -332,11 +333,25 @@ proc validateAndRelay(g: GossipSub,
...
     # Don't send it to source peer, or peers that
     # sent it during validation
     toSendPeers.excl(peer)
     toSendPeers.excl(seenPeers)
 
+    if msg.data.len > msgId.len * 10:
```
Comment (diegomrsantos) [inline libp2p/protocols/pubsub/gossipsub.nim:342]: Can we make this condition more descriptive - like making it a proc or adding a comment with the reasoning?

File: libp2p/protocols/pubsub/gossipsub/behavior.nim [hunk 3]
```diff
@@ -262,6 +262,15 @@ proc handleIHave*(g: GossipSub,
     g.rng.shuffle(res.messageIds)
     return res
 
+proc handleIDontWant*(g: GossipSub,
+                      peer: PubSubPeer,
+                      iDontWants: seq[ControlIWant]) =
+  for dontWant in iDontWants:
+    for message in dontWant.messageIds:
```
Comment (diegomrsantos) [inline libp2p/protocols/pubsub/gossipsub/behavior.nim:269]: `messageId`?

File: libp2p/protocols/pubsub/gossipsub/behavior.nim [hunk 4]
```diff
@@ -262,6 +262,15 @@ proc handleIHave*(g: GossipSub,
...
+      if peer.heDontWants[^1].len > 1000: break
+      if message.len > 100: break
```
Comment (diegomrsantos) [inline libp2p/protocols/pubsub/gossipsub/behavior.nim:271]: should it be a `continue`?

File: libp2p/protocols/pubsub/pubsubpeer.nim [hunk 5]
```diff
@@ -60,6 +60,7 @@ type
 
     score*: float64
     sentIHaves*: Deque[HashSet[MessageId]]
+    heDontWants*: Deque[HashSet[MessageId]]
```
[the remark this answers is not in the document]
Comment (Menduist) [reply libp2p/protocols/pubsub/pubsubpeer.nim:63]: This is the list of "iDontWant" he sent, so I think "heDontWants" or "receivedDontWants" is more descriptive

[the remark this answers is not in the document]
Comment (diegomrsantos) [reply libp2p/protocols/pubsub/pubsubpeer.nim:63]: so maybe `rcvDontWants`.

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 6]
```diff
@@ -337,6 +338,21 @@ proc validateAndRelay(g: GossipSub,
     toSendPeers.excl(peer)
     toSendPeers.excl(seenPeers)
 
+    # IDontWant is only worth it if the message is substantially
+    # bigger than the messageId
+    if msg.data.len > msgId.len * 10:
+      g.broadcast(toSendPeers, RPCMsg(control: some(ControlMessage(
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/gossipsub.nim:344]: we should only broadcast this to peers that haven't themselves sent us an idontwant for the given hash

File: libp2p/protocols/pubsub/gossipsub/behavior.nim [hunk 7]
```diff
@@ -629,6 +638,9 @@ proc onHeartbeat(g: GossipSub) {.raises: [].} =
         peer.sentIHaves.addFirst(default(HashSet[MessageId]))
         if peer.sentIHaves.len > g.parameters.historyLength:
           discard peer.sentIHaves.popLast()
+        peer.heDontWants.addFirst(default(HashSet[MessageId]))
+        if peer.heDontWants.len > g.parameters.historyLength:
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/gossipsub/behavior.nim:642]: `while`?

File: libp2p/protocols/pubsub/pubsubpeer.nim (hunk 5 above)
[the remark this answers is not in the document]
Comment (arnetheduck) [reply libp2p/protocols/pubsub/pubsubpeer.nim:63]: just `dontWants` is enough really

**Pull request merged**